### PR TITLE
planner: fix plan cache rebuild range error

### LINKED ISSUE #42150: runtime error: index out of range [0] with length 0
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```

use test;
drop table if exists t1, t2;
CREATE TABLE `t1` (
  `c_int` int(11) NOT NULL,
  `c_str` varchar(40) CHARACTER SET utf8 COLLATE utf8_bin DEFAULT NULL,
  `c_datetime` datetime DEFAULT NULL,
  `c_timestamp` timestamp NULL DEFAULT NULL,
  `c_double` double DEFAULT NULL,
  `c_decimal` decimal(12,6) DEFAULT NULL,
  `c_enum` enum('blue','green','red','yellow','white','orange','purple') NOT NULL,
  PRIMARY KEY (`c_int`,`c_enum`) /*T![clustered_index] CLUSTERED */,
  KEY `c_decimal` (`c_decimal`),
  UNIQUE KEY `c_datetime` (`c_datetime`),
  UNIQUE KEY `c_timestamp` (`c_timestamp`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
CREATE TABLE `t2` (
  `c_int` int(11) NOT NULL,
  `c_str` varchar(40) CHARACTER SET utf8 COLLATE utf8_bin DEFAULT NULL,
  `c_datetime` datetime DEFAULT NULL,
  `c_timestamp` timestamp NULL DEFAULT NULL,
  `c_double` double DEFAULT NULL,
  `c_decimal` decimal(12,6) DEFAULT NULL,
  `c_enum` enum('blue','green','red','yellow','white','orange','purple') NOT NULL,
  PRIMARY KEY (`c_int`,`c_enum`) /*T![clustered_index] CLUSTERED */,
  KEY `c_decimal` (`c_decimal`),
  UNIQUE KEY `c_datetime` (`c_datetime`),
  UNIQUE KEY `c_timestamp` (`c_timestamp`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;

/* test */ /* t02:008 */ prepare stmt16 from "select * from t1 where c_enum in (select c_enum from t2 where t1.c_str in (?, ?, ?) )  ";
/* test */ /* t02:008 */ set @v0 = "nice hellman";
/* test */ /* t02:008 */ set @v1 = "flamboyant booth";
/* test */ /* t02:008 */ set @v2 = "quirky brahmagupta";
/* test:query */ /* t02:008 */ execute stmt16 using @v0, @v1, @v2;
/* test:query */ /* t02:008 */ execute stmt16 using @v0, @v1, @v2;
```
### 2. What did you expect to see? (Required)
```
MySQL [test]> /* test:query */ /* t02:008 */ execute stmt16 using @v0, @v1, @v2;
Empty set (0.01 sec)

MySQL [test]> /* test:query */ /* t02:008 */ execute stmt16 using @v0, @v1, @v2;
Empty set (0.01 sec)
```
### 3. What did you see instead (Required)
```
MySQL [test]> /* test:query */ /* t02:008 */ execute stmt16 using @v0, @v1, @v2;
Empty set (0.01 sec)

MySQL [test]> /* test:query */ /* t02:008 */ execute stmt16 using @v0, @v1, @v2;
ERROR 1105 (HY000): runtime error: index out of range [0] with length 0
```
### 4. What is your TiDB version? (Required)

```
| Release Version: v6.7.0-alpha-246-g145b7cd
Edition: Community
Git Commit Hash: 145b7cdf72feeecf7f2b0b3e01693c603712dce5
Git Branch: master
UTC Build Time: 2023-03-13 01:04:19
GoVersion: go1.20.1
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: unistore |
```

### PR DESCRIPTION
**Pull request by fzzf678**
### What problem does this PR solve?

Issue Number: close #42150

Problem Summary:
When rebuild range for `IndexFullScan`, if there is no filter on columns, the type slice may out of bound. There should be early stop logic facing this situation.
### What is changed and how it works?
Return full range when build index full scan without filter cols
### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/plan_cache_test.go [hunk 1]
```diff
@@ -1443,3 +1443,22 @@ func TestNonPreparedPlanExplainWarning(t *testing.T) {
 		require.Equal(t, reasons[idx], warn[2])
 	}
 }
+
+func TestIssue42150(t *testing.T) {
```
Comment (qw4990) [inline planner/core/plan_cache_test.go:1447]: Please add this minimum case into this test as well:
```
mysql> create table t (a int, b int, primary key(a), key(b));
Query OK, 0 rows affected (0.03 sec)

mysql> prepare st from 'select a from t use index(b)';
Query OK, 0 rows affected (0.00 sec)

mysql> execute st;
Empty set (0.00 sec)

mysql> execute st;
ERROR 1105 (HY000): runtime error: index out of range [0] with length 0
```

File: planner/core/plan_cache.go [hunk 2]
```diff
@@ -662,6 +662,10 @@ func buildRangeForTableScan(sctx sessionctx.Context, ts *PhysicalTableScan) (err
 }
 
 func buildRangeForIndexScan(sctx sessionctx.Context, is *PhysicalIndexScan) (err error) {
+	if len(is.IdxCols) == 0 { // return full range when PhysicalIndexScan has no IdxCols
+		is.Ranges = ranger.FullRange()
```
Comment (qw4990) [inline planner/core/plan_cache.go:666]:
```
if IsFullRagne(is.Ranges) { // the original range is already a full-range.
    return nil
} else {
    return error("unexpected")
}
```

Comment (xuyifangreeneyes) [reply planner/core/plan_cache.go:666]: @qw4990 Do you mean that we should assert `IsFullRagne(is.Ranges)` must stand when `len(is.IdxCols) == 0`?

Comment (qw4990) [reply planner/core/plan_cache.go:666]: Yes, I'm not sure whether `len(is.IdxCols)==0` can guarantee this is a `IndexFullScan`, so check `IsFullRagne(is.Ranges)` for safety.

Comment (fzzf678) [reply planner/core/plan_cache.go:666]: got it

**Comment by fzzf678** [comment]
@qw4990 @xuyifangreeneyes updated, TPAL again

**Review by qw4990 — APPROVED** [review]

**Review by xuyifangreeneyes — APPROVED** [review]

**Comment by fzzf678** [comment]
this bug won't affect 5.4 6.1 6.5, but we need cherry-pick a test to these version for safety

**Pull request merged**